{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GRPM Dataset Builder 2.0 (LitVar2 Patch)\n",
    "\n",
    "This notebook is designed to retrieve genetic polymorphism data from multiple sources. It uses the LitVar API to extract polymorphisms for each human gene within the LitVar2 database along with all associated PubMed Identifiers (PMIDs). These PMIDs are then employed as queries on PubMed to obtain MEDLINE data (parsed though the 'nbib' package). All collected data are ultimately consolidated into a single CSV file, known as the \"GRPM Dataset\", which serves as the primary source against which MeSH term queries can be launched to retrieve genes and polymorphisms associated with specific contexts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-03-14T10:08:27.549760300Z",
     "start_time": "2024-03-14T10:08:27.516843900Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current working directory: G:\\Altri computer\\Horizon\\horizon_workspace\\projects\\bioinformatics\\semantics\\GRPM\\GRPM_dev\n"
     ]
    }
   ],
   "source": [
    "import ast\n",
    "#Only for Google Colab\n",
    "import os\n",
    "import sys\n",
    "\n",
    "# @markdown Run in Colab virtual machine by default\n",
    "\n",
    "# @markdown to run in google drive set:\n",
    "import_mydrive = False #@param {type:\"boolean\"}\n",
    "\n",
    "if 'google.colab' in sys.modules:\n",
    "    !pip install nbib\n",
    "    !pip install biopython\n",
    "\n",
    "    if import_mydrive:\n",
    "        from google.colab import drive\n",
    "        drive.mount('/content/drive')\n",
    "        if os.path.exists('/content/drive/MyDrive/grpm_system/'):\n",
    "            %cd /content/drive/MyDrive/grpm_system/\n",
    "        else:\n",
    "            %mkdir /content/drive/MyDrive/grpm_system/\n",
    "            %cd /content/drive/MyDrive/grpm_system/\n",
    "    else:\n",
    "        if os.path.exists('/content/grpm_system/'):\n",
    "            %cd /content/grpm_system/\n",
    "        else:\n",
    "            %mkdir /content/grpm_system/\n",
    "            %cd /content/grpm_system/\n",
    "\n",
    "current_directory = os.getcwd()\n",
    "print(\"Current working directory:\", current_directory)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get required data from Zenodo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-03-14T10:08:44.511914900Z",
     "start_time": "2024-03-14T10:08:41.231007900Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ZIP file 'human_genes_repo.zip' extracted to 'G:\\Altri computer\\Horizon\\horizon_workspace\\projects\\bioinformatics\\semantics\\GRPM\\GRPM_dev' successfully.\n"
     ]
    }
   ],
   "source": [
    "# Get Required datasets from Zenodo Repository\n",
    "#https://zenodo.org/record/8205724  DOI: 10.5281/zenodo.8205724\n",
    "\n",
    "import os\n",
    "import io\n",
    "import requests\n",
    "import zipfile\n",
    "\n",
    "def get_and_extract(file, dir = os.getcwd()):\n",
    "    url = \"https://zenodo.org/record/8205724/files/\"+file+\".zip?download=1\"\n",
    "    zip_file_name = file+\".zip\"\n",
    "    extracted_folder_name = dir\n",
    "\n",
    "    # Download the ZIP file\n",
    "    response = requests.get(url)\n",
    "\n",
    "    if response.status_code == 200:\n",
    "        # Extract the ZIP contents\n",
    "        with io.BytesIO(response.content) as zip_buffer:\n",
    "            with zipfile.ZipFile(zip_buffer, 'r') as zip_ref:\n",
    "                zip_ref.extractall(extracted_folder_name)\n",
    "        print(f\"ZIP file '{zip_file_name}' extracted to '{extracted_folder_name}' successfully.\")\n",
    "    else:\n",
    "        print(\"Failed to download the ZIP file.\")\n",
    "\n",
    "if not os.path.exists('human_geness_repo'):\n",
    "    get_and_extract('human_genes_repo')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-03-14T10:09:57.491229200Z",
     "start_time": "2024-03-14T10:09:54.585650300Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The module 'nbib' is already installed.\n",
      "The module 'requests' is already installed.\n"
     ]
    }
   ],
   "source": [
    "import importlib\n",
    "\n",
    "def simple_bool(message):\n",
    "    choose = input(message+\" (y/n): \").lower()\n",
    "    your_bool = choose in [\"y\", \"yes\"]\n",
    "    return your_bool\n",
    "\n",
    "def check_and_install_module(module_name):\n",
    "    try:\n",
    "        # Check if the module is already installed\n",
    "        importlib.import_module(module_name)\n",
    "        print(f\"The module '{module_name}' is already installed.\")\n",
    "    except ImportError:\n",
    "        # If the module is not installed, try installing it\n",
    "        x = simple_bool(\n",
    "            \"\\n\" + module_name + \"  module is not installed.\\nwould you like to install it?\")\n",
    "        if x:\n",
    "            import subprocess\n",
    "            subprocess.check_call([\"pip\", \"install\", module_name])\n",
    "            print(f\"The module '{module_name}' was installed correctly.\")\n",
    "        else:\n",
    "            pass\n",
    "\n",
    "check_and_install_module('nbib')\n",
    "check_and_install_module('requests')\n",
    "\n",
    "#Import Modules\n",
    "import os\n",
    "import sys\n",
    "import json\n",
    "import nbib\n",
    "import pandas as pd\n",
    "import requests as rq\n",
    "import matplotlib.pyplot as plt\n",
    "from datetime import datetime\n",
    "from bs4 import BeautifulSoup\n",
    "from io import StringIO\n",
    "from Bio import Entrez\n",
    "import time\n",
    "\n",
    "Entrez.email = \"your_email@example.com\"\n",
    "\n",
    "request_counter = 0\n",
    "gene_counter = 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Whole genome forecast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-03-14T10:10:03.870275900Z",
     "start_time": "2024-03-14T10:10:03.737634700Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Forecast:\n",
      "max genes/day=  2352\n",
      "for 19383 genes:\n",
      "    request counter = 82377.75 requests\n",
      "    whole genome runtime = 34.24 hours\n",
      "    whole genome runtime overnight = 40.17\n",
      "    db table size 7695.05 MB\n"
     ]
    }
   ],
   "source": [
    "#Statistics based on 150 random genes:\n",
    "time_sleep = 0.4\n",
    "runtime_gene = 6.36 #sec/gene\n",
    "genes_hour = 566 #genes/hour\n",
    "request_counter_gene = 4.25 #request/gene (with base sleep (0.4))\n",
    "sleep_request_base = 0.4 #time sleep each request\n",
    "sleep_request_overnight_plus = 1.1 # for an overnight job\n",
    "\n",
    "print('Forecast:')\n",
    "max_genes = int(10000/request_counter_gene)\n",
    "table_size_db_gene = 0.496 #MB\n",
    "table_size_gene = 0.397 #MB\n",
    "png_size_db_gene = 0.47 #KB\n",
    "\n",
    "#Forecast:\n",
    "genes = pd.read_csv('human_genes_repo/H_GENES_proteincoding_genes.csv')\n",
    "ngenes = len(genes)#gene_range\n",
    "nruntime = ngenes * runtime_gene\n",
    "#print('runtime, '+str(ngenes), nruntime)\n",
    "nrequest_counter = ngenes * request_counter_gene\n",
    "\n",
    "tempo_ore = round(nruntime/3600, 2)\n",
    "tempo_ore_overnight = round((nruntime+(sleep_request_overnight_plus*ngenes))/3600, 2)\n",
    "\n",
    "print('max genes/day= ',max_genes)\n",
    "print('for',str(int(ngenes)),'genes:')\n",
    "print('    request counter =', nrequest_counter,'requests')\n",
    "print('    whole genome runtime =', tempo_ore,'hours')\n",
    "print('    whole genome runtime overnight =', tempo_ore_overnight)\n",
    "\n",
    "db_table_size = ngenes * table_size_gene\n",
    "print('    db table size', round(db_table_size,2),'MB')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Human Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-03-14T10:10:06.104328600Z",
     "start_time": "2024-03-14T10:10:05.922359200Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "protein_coding_genes 19318 \n",
      "IG_TR_genes 641 \n",
      "RNA_genes 11452 \n",
      "pseudo_genes 9866 \n",
      "misc_genes 22\n",
      "\n",
      "recommended job lenght for pcg: 1073\n"
     ]
    }
   ],
   "source": [
    "# Load Human Gene list ---------------------------------\n",
    "protein_coding_genes = pd.read_csv('human_genes_repo/H_GENES_proteincoding_genes.csv')\n",
    "IG_TR_genes          = pd.read_csv('human_genes_repo/H_GENES_IGTR_genes.csv')\n",
    "RNA_genes            = pd.read_csv('human_genes_repo/H_GENES_RNA_genes.csv')\n",
    "pseudo_genes         = pd.read_csv('human_genes_repo/H_GENES_pseudo_genes.csv')\n",
    "misc_genes           = pd.read_csv('human_genes_repo/H_GENES_misc_genes.csv')\n",
    "\n",
    "# create gene lists:\n",
    "protein_coding_genes_list = protein_coding_genes['Gene name'].dropna().tolist()\n",
    "rna_genes_list = RNA_genes['Gene name'].dropna().tolist()\n",
    "pseudo_genes_list = pseudo_genes['Gene name'].dropna().tolist()\n",
    "\n",
    "\n",
    "# Split job packages:----------------------------------\n",
    "\n",
    "# (1) protein coding genes:\n",
    "gene_range = int(len(protein_coding_genes_list)/18)\n",
    "genes = [protein_coding_genes_list[i * gene_range : (i + 1) * gene_range] for i in range(0, 18)]\n",
    "pcg_chunks = genes[:18]\n",
    "\n",
    "# (2) RNA genes:\n",
    "rna_gene_range = int(len(rna_genes_list)/5)\n",
    "genes = [rna_genes_list[i * rna_gene_range : (i + 1) * rna_gene_range] for i in range(0, 8)]\n",
    "rna_chunks = genes[:5]\n",
    "\n",
    "# (3) pseudo genes:\n",
    "pseudo_gene_range = int(len(pseudo_genes_list)/2)\n",
    "genes = [rna_genes_list[i * pseudo_gene_range : (i + 1) * pseudo_gene_range] for i in range(0, 8)]\n",
    "pseudo_chunks = genes[:2]\n",
    "\n",
    "print('protein_coding_genes',len(protein_coding_genes['Gene name'].dropna()),\n",
    "      '\\nIG_TR_genes',len(IG_TR_genes['Gene name'].dropna()),\n",
    "      '\\nRNA_genes',len(RNA_genes['Gene name'].dropna()),\n",
    "      '\\npseudo_genes',len(pseudo_genes['Gene name'].dropna()),\n",
    "      '\\nmisc_genes',len(misc_genes['Gene name'].dropna()))\n",
    "\n",
    "print('\\nrecommended job lenght for pcg:',int(len(protein_coding_genes_list)/18))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set options and import building dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-03-14T11:12:19.023648700Z",
     "start_time": "2024-03-14T11:12:18.981945900Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "time load 0:00:00.003002\n",
      "empty dataset\n",
      "empty dataset\n"
     ]
    }
   ],
   "source": [
    "# set options:--------------------------------\n",
    "save_plot           = False\n",
    "save_studytype_data = False\n",
    "save_accessory_data = False\n",
    "\n",
    "# set db:--------------------------------\n",
    "db_tag = 'pcg_2024_lit2'\n",
    "    # 'pcg'    = protein coding genes = grpm_db\n",
    "    # 'rna'    = rna genes            = grpm_db_rna\n",
    "    # 'pseudo' = pseudogenes          = grpm_db_pseudo\n",
    "\n",
    "db_name = 'grpm_db_'+db_tag\n",
    "db_path = 'grpm_dataset/'+db_name\n",
    "\n",
    "if not os.path.exists(db_path):\n",
    "    os.makedirs(db_path)\n",
    "#------------------------------------------------\n",
    "\n",
    "# set gene job checkpoint frequency\n",
    "checkpoint = 3\n",
    "\n",
    "#import checkpoint datasets:\n",
    "time_a = datetime.now()\n",
    "if os.path.isfile(db_path+'/grpm_table_output.csv'):\n",
    "    complete_df = pd.read_csv(db_path+'/grpm_table_output.csv',index_col=0)\n",
    "    restart = True\n",
    "else:\n",
    "    complete_df = pd.DataFrame()\n",
    "    restart = False\n",
    "\n",
    "if os.path.isfile(db_path+'/complete_nbibtable.csv'):\n",
    "    complete_nbibtable = pd.read_csv(db_path+'/complete_nbibtable.csv',index_col=0)\n",
    "else:\n",
    "    complete_nbibtable = pd.DataFrame()\n",
    "time_b = datetime.now()\n",
    "\n",
    "print('time load',time_b-time_a)\n",
    "\n",
    "## check saved data:\n",
    "if os.path.isfile(db_path+'/grpm_table_output.csv'):\n",
    "    gene_db_count =  complete_df.gene.nunique()\n",
    "    print('complete_df gene count:',gene_db_count,'on', len(protein_coding_genes_list))\n",
    "    if gene_db_count >= 15519:\n",
    "        print('grpm db already contains all available genes on litvar1')\n",
    "\n",
    "    print('\\ngrpm_table_output.csv size'  ,round(os.path.getsize(db_path+'/grpm_table_output.csv')/(1024*1024),3),'MB')\n",
    "    print('complete_nbibtable.csv size',round(os.path.getsize(db_path+'/complete_nbibtable.csv')/(1024*1024),3),'MB')\n",
    "    print('memory_usage_complete_df'     ,round(complete_df.memory_usage().sum()/(1024*1024),3))\n",
    "    print('memory_usage_complete_nbib_df',round(complete_nbibtable.memory_usage().sum()/(1024*1024),3))\n",
    "else:\n",
    "    print('empty dataset')\n",
    "    print('empty dataset')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run Job"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set gene-range for this job"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-03-14T11:12:40.990520500Z",
     "start_time": "2024-03-14T11:12:22.900297500Z"
    }
   },
   "outputs": [],
   "source": [
    "# Set the gene list ----------------\n",
    "\n",
    "#1. Set gene-range [whole genome build]\n",
    "    # pcg_chunks    [0:17]\n",
    "    # rna_chunks    [0:4]\n",
    "    # pseudo_chunks [0:1]\n",
    "gene_chunk = pcg_chunks[0]\n",
    "\n",
    "#2. place here your custom gene list [custom build]\n",
    "custom_genes = ['APOA1', 'FFC1', 'ERH', 'USP53']\n",
    "custom_list = simple_bool('Do you want to run the custom list?')\n",
    "\n",
    "#if stucked, store skipped_genes for later:\n",
    "skipped_genes =  []\n",
    "\n",
    "# run a sample?\n",
    "run_sample = simple_bool('Do you want to run a sample?')\n",
    "if run_sample:\n",
    "    sample_size = int(input('sample size? \\nnum:'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Job"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-03-14T11:19:50.089650Z",
     "start_time": "2024-03-14T11:18:02.706681900Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start at  2024-03-14 12:18:02.771662\n",
      "ATP23_runtime: 0:00:04.451173\n",
      "CENATAC_runtime: 0:00:04.992364\n",
      "ZFY_runtime: 0:00:04.429840\n",
      "saved checkpoint\n",
      "NLRP2_runtime: 0:00:05.577129\n",
      "ISX_runtime: 0:00:04.555840\n",
      "ANKRD12_runtime: 0:00:04.969481\n",
      "saved checkpoint\n",
      "SFI1_runtime: 0:00:06.460779\n",
      "STC1_runtime: 0:00:04.654620\n",
      "NEU1_runtime: 0:00:10.444339\n",
      "saved checkpoint\n",
      "CCL3L3_runtime: 0:00:04.824262\n",
      "GSTT4_runtime: 0:00:04.620403\n",
      "GSTT2_runtime: 0:00:04.489459\n",
      "saved checkpoint\n",
      "ZC3H3_runtime: 0:00:05.588509\n",
      "PRAMEF2_runtime: 0:00:04.663845\n",
      "ADAM18_runtime: 0:00:04.809382\n",
      "saved checkpoint\n",
      "ZBED4_runtime: 0:00:05.242171\n",
      "PRPF31_runtime: 0:00:06.089593\n",
      "MMEL1_runtime: 0:00:05.563567\n",
      "saved checkpoint\n",
      "NPIPA7_runtime: 0:00:04.321967\n",
      "GGNBP2_runtime: 0:00:05.493570\n",
      "gene batch: 20\n",
      "time batch: 0:01:47\n",
      "runtime/gene: 0:00:05.365560\n",
      "request_counter: 117  (limit: 10.000/day)\n",
      "requests/gene: 3.25\n",
      "2024-03-14 12:19:50.082866\n"
     ]
    }
   ],
   "source": [
    "import ast\n",
    "# Run Job -------------------------\n",
    "\n",
    "if run_sample:\n",
    "    genes = pd.Series(gene_chunk).sample(sample_size).reset_index(drop=True)\n",
    "    genes = genes.to_list()\n",
    "    restart = False\n",
    "else:\n",
    "    genes = pd.Series(gene_chunk)\n",
    "    genes = genes.to_list()\n",
    "\n",
    "if restart:\n",
    "    restart_from = complete_df.gene.nunique()\n",
    "    gene_start = restart_from\n",
    "    print('search restarted from '+str(restart_from))\n",
    "else:\n",
    "    gene_start = 0\n",
    "\n",
    "if custom_list:\n",
    "    genes = custom_genes\n",
    "    gene_start = 0\n",
    "\n",
    "time_start = datetime.now()\n",
    "print('Start at ',time_start)\n",
    "\n",
    "# get gene data\n",
    "for gene in genes[gene_start:]:\n",
    "\n",
    "    #LitVar2 \"Variants for Gene\" API request\n",
    "    if request_counter > 9950:\n",
    "        print('Request limit reached. Wait \\'till tomorrow!')\n",
    "        pass\n",
    "    time_alpha = datetime.now()\n",
    "    url = \"https://www.ncbi.nlm.nih.gov/research/litvar2-api/variant/search/gene/\" + gene\n",
    "    respose = (rq.get(url)).text\n",
    "\n",
    "    # parsing output in JSON\n",
    "    def transform_string(foo):\n",
    "        foo = foo.replace(\"\\n\",\", \")\n",
    "        foo = foo.replace(\"\\'\",\"\\\"\")\n",
    "        foo = foo.replace('\\\"\\\"', '\\\"')\n",
    "        foo = foo.replace('p.\\\"','p.')\n",
    "        foo = foo.replace('c.\\\"','c.')\n",
    "        foo = foo.replace('g.\\\"','g.')\n",
    "        foo = foo.replace('\\\">','>')\n",
    "        foo = foo.replace('.C\\\"204','.C204') #<= if stucked, look for bugs like this into text\n",
    "        return foo\n",
    "    \n",
    "    data= \"[\" + transform_string(respose) + \"]\"\n",
    "    \n",
    "    #Create Dataframe\n",
    "    df = pd.read_json(StringIO(data))\n",
    "    if 'rsid' in df.columns and len(df.rsid)>1:\n",
    "        # creare un df senza i clingen.\n",
    "        dfb = df[['_id','pmids_count','rsid']]\n",
    "        dfa = dfb[~dfb['_id'].str.contains('@CA')].drop_duplicates().reset_index(drop=True)\n",
    "        dfn = dfa.dropna(subset=['rsid'])\n",
    "\n",
    "        #Statistics\n",
    "        #handle = Entrez.esearch(db=\"snp\", term=gene)\n",
    "        #record = Entrez.read(handle)\n",
    "        #request_counter += 1\n",
    "\n",
    "        NCBI_dbSNP = 'na' #record[\"Count\"]\n",
    "        lit2_variant = len(dfa['_id'].drop_duplicates())\n",
    "        lit2_variant_norsid = len(dfa.loc[df['rsid'].isna()])\n",
    "        lit2_rsid = len(dfn.rsid.drop_duplicates())\n",
    "\n",
    "\n",
    "        # remove rs with pmid_count = 1\n",
    "        df2 = dfn.loc[df.pmids_count !=1]#.reset_index(drop=True)\n",
    "        lit2_rsid_f = len(df2)\n",
    "\n",
    "        # accessory data\n",
    "        dfsort = df.sort_values(by='pmids_count',ascending=False).reset_index(drop=True)\n",
    "        df2sort = df2.sort_values(by='pmids_count',ascending=False).reset_index(drop=True)\n",
    "\n",
    "        r'''#LitVar1 POST rsid2pmids\n",
    "        # convert in \"data\" for POST request\n",
    "        liss = list(map(str, df.rsid))\n",
    "        qrsid = \"\\\", \\\"\".join(liss)\n",
    "        qrsid = '{\"rsids\": [\"'+ qrsid +'\"]}'\n",
    "        \n",
    "        \n",
    "        url=\"https://www.ncbi.nlm.nih.gov/research/bionlp/litvar/api/v1/public/rsids2pmids\"\n",
    "        headers={ \"Content-Type\" : \"application/json\", \"Accept\" : \"application/json\"}\n",
    "\n",
    "        # enter my data = rsID list\n",
    "        data = qrsid\n",
    "        #data = {[\"rs781028867\", \"rs17817449\"]}\n",
    "        r = rq.post(url, headers=headers, data=data )\n",
    "        if not r.ok:\n",
    "            r.raise_for_status()\n",
    "            sys.exit()\n",
    "        rspost = r.json()'''\n",
    "        \n",
    "\n",
    "        for rsid in df2.rsid[0:1]:\n",
    "            url=\"https://www.ncbi.nlm.nih.gov/research/litvar2-api/variant/get/litvar@\"+rsid+\"%23%23/publications\"\n",
    "            #print(url)\n",
    "            respose = (rq.get(url)).text\n",
    "            respose_dict = ast.literal_eval(respose)\n",
    "            #print(respose_dict['pmids'])\n",
    "            rspost = pd.DataFrame(respose_dict)\n",
    "\n",
    "            #Display-------------------------------------------------------\n",
    "            dfrspost = pd.DataFrame(rspost)\n",
    "            if 'pmids' in dfrspost.columns and len(dfrspost.pmids_count)>1:\n",
    "            #if 'rsid' in dfrspost.columns and len(dfrspost.rsid)>1:\n",
    "                #lit1_rsid = len(dfrspost.rsid)\n",
    "                lit1_rsid = 0\n",
    "                #lit2_rsid = len(df)\n",
    "    \n",
    "                # Creating the simple list [rsid-pmid]========================\n",
    "                #[MODULE: \"Alligner\"]\n",
    "                #rspmid = []\n",
    "                #for i in range(len(dfrspost)):\n",
    "                #    for pmid in dfrspost['pmids'][i]: #dfrspost = mother table\n",
    "                #        out = dfrspost['rsid'][i], pmid\n",
    "                #        rspmid.append(out)\n",
    "                \n",
    "    \n",
    "                #rsidpmid = pd.DataFrame(rspmid).drop_duplicates().rename(columns={0: 'rsid',1:'pmids'})\n",
    "                #rsidpmid['pmids'] = rsidpmid['pmids'].astype(str) \n",
    "    \n",
    "                rsidpmid = dfrspost[['pmids']].copy()\n",
    "                rsidpmid['rsid'] = str(rsid)\n",
    "                rsidpmid['pmids'] = rsidpmid['pmids'].astype('str')\n",
    "\n",
    "                \n",
    "                #report data:\n",
    "                lit1_rsid_pmid = len(rsidpmid)\n",
    "                lit1_pmid = len(rsidpmid.drop_duplicates(subset='pmids'))\n",
    "    \n",
    "    \n",
    "                ####[MODULE: groupby.describe]\n",
    "                # applicare groupby ad rsidpmid per avere tabella pmid count\n",
    "                rsidpmidcount = rsidpmid.groupby('rsid').describe().reset_index()\n",
    "                rsidpmidcount.columns = rsidpmidcount.columns.to_flat_index()\n",
    "            \n",
    "\n",
    "                #replace column names\n",
    "                new_column_names = ['rsid', 'pmid_count', 'pmid_unique','pmid_top','pmid_freq']\n",
    "                rsidpmidcount.columns = new_column_names\n",
    "                rsidpmidcountf = rsidpmidcount[['rsid','pmid_unique']]\n",
    "    \n",
    "                #report data:-------------------------------------------------------------------------\n",
    "                lit1_rsid_f = len(rsidpmidcountf[rsidpmidcountf.pmid_unique!=1])\n",
    "                lit1_rsid_m = len(rsidpmidcountf[rsidpmidcountf.pmid_unique==1])\n",
    "    \n",
    "                rsidpmidcountfsort = rsidpmidcountf.sort_values('pmid_unique',ascending=False).reset_index(drop=True)\n",
    "    \n",
    "    \n",
    "                #Filter pmid for rsid with pmid>1-----------------------------------------------------\n",
    "                outless = rsidpmidcountfsort[rsidpmidcountfsort.pmid_unique>1]\n",
    "                #creare una mask isin su rsidpmid con outless.rsid\n",
    "                mask = rsidpmid['rsid'].isin(outless.rsid)\n",
    "                rsidpmidless = rsidpmid[mask]\n",
    "                lit1_pmid_f = len(rsidpmidless.pmids.drop_duplicates())\n",
    "    \n",
    "    \n",
    "                # PubMed query Build:=====================================================\n",
    "                ### two input alternatives (total LitVar1 and LitVar>1)\n",
    "                #Total\n",
    "                pmid_l = rsidpmid.pmids.drop_duplicates().tolist()\n",
    "                \n",
    "                #Query chunk build (max:1300)\n",
    "                limit = 1300\n",
    "                \n",
    "                ##Define list of queries for PubMed:\n",
    "                query = []\n",
    "                \n",
    "                def query_build(pmid_list):\n",
    "                    query = \"+OR+\".join(pmid_list)\n",
    "                    return query\n",
    "                \n",
    "                if len(pmid_l)<=limit:\n",
    "                    pmid_l01 = pmid_l\n",
    "                    query = [query_build(pmid_l01)]\n",
    "                \n",
    "                if limit<len(pmid_l)<=limit*2:\n",
    "                    j = len(pmid_l)//2\n",
    "                    pmid_l01 = pmid_l[:j]\n",
    "                    pmid_l02 = pmid_l[j:]\n",
    "                    query = [query_build(pmid_l01),\n",
    "                             query_build(pmid_l02)]\n",
    "                \n",
    "                if limit*2<len(pmid_l)<=limit*3:\n",
    "                    j = len(pmid_l)//3\n",
    "                    pmid_l01 = pmid_l[:j]\n",
    "                    pmid_l02 = pmid_l[j:j*2]\n",
    "                    pmid_l03 = pmid_l[j*2:]\n",
    "                    query = [query_build(pmid_l01),\n",
    "                             query_build(pmid_l02),\n",
    "                             query_build(pmid_l03)]\n",
    "                \n",
    "                if limit*3<len(pmid_l)<=limit*4:\n",
    "                    j = len(pmid_l)//4\n",
    "                    pmid_l01 = pmid_l[:j]\n",
    "                    pmid_l02 = pmid_l[j:j*2]\n",
    "                    pmid_l03 = pmid_l[j*2:j*3]\n",
    "                    pmid_l04 = pmid_l[j*3:]\n",
    "                    query = [query_build(pmid_l01),\n",
    "                             query_build(pmid_l02),\n",
    "                             query_build(pmid_l03),\n",
    "                             query_build(pmid_l04)]\n",
    "                \n",
    "                if limit*4<len(pmid_l)<=limit*5:\n",
    "                    j = len(pmid_l)//5\n",
    "                    pmid_l01 = pmid_l[:j]\n",
    "                    pmid_l02 = pmid_l[j:j*2]\n",
    "                    pmid_l03 = pmid_l[j*2:j*3]\n",
    "                    pmid_l04 = pmid_l[j*3:j*4]\n",
    "                    pmid_l05 = pmid_l[j*4:]\n",
    "                    query = [query_build(pmid_l01),\n",
    "                             query_build(pmid_l02),\n",
    "                             query_build(pmid_l03),\n",
    "                             query_build(pmid_l04),\n",
    "                             query_build(pmid_l05)]\n",
    "                \n",
    "                if limit*5<len(pmid_l)<=limit*6:\n",
    "                    j = len(pmid_l)//6\n",
    "                    pmid_l01 = pmid_l[:j]\n",
    "                    pmid_l02 = pmid_l[j:j*2]\n",
    "                    pmid_l03 = pmid_l[j*2:j*3]\n",
    "                    pmid_l04 = pmid_l[j*3:j*4]\n",
    "                    pmid_l05 = pmid_l[j*4:j*5]\n",
    "                    pmid_l06 = pmid_l[j*5:]\n",
    "                    query = [query_build(pmid_l01),\n",
    "                             query_build(pmid_l02),\n",
    "                             query_build(pmid_l03),\n",
    "                             query_build(pmid_l04),\n",
    "                             query_build(pmid_l05),\n",
    "                             query_build(pmid_l06)]\n",
    "                \n",
    "                if limit*6<len(pmid_l)<=limit*7:\n",
    "                    j = len(pmid_l)//7\n",
    "                    pmid_l01 = pmid_l[:j]\n",
    "                    pmid_l02 = pmid_l[j:j*2]\n",
    "                    pmid_l03 = pmid_l[j*2:j*3]\n",
    "                    pmid_l04 = pmid_l[j*3:j*4]\n",
    "                    pmid_l05 = pmid_l[j*4:j*5]\n",
    "                    pmid_l06 = pmid_l[j*5:j*6]\n",
    "                    pmid_l07 = pmid_l[j*6:]\n",
    "                    query = [query_build(pmid_l01),\n",
    "                             query_build(pmid_l02),\n",
    "                             query_build(pmid_l03),\n",
    "                             query_build(pmid_l04),\n",
    "                             query_build(pmid_l05),\n",
    "                             query_build(pmid_l06),\n",
    "                             query_build(pmid_l07)]\n",
    "    \n",
    "    \n",
    "    \n",
    "        # Merging requests for the queries\n",
    "                ### carefull: high runtime\n",
    "    \n",
    "                time1 = datetime.now()\n",
    "                pages = ((len(pmid_l01)//200)+1)+1\n",
    "                if len(pmid_l01) % 200 == 0:\n",
    "                    pages = pages -1\n",
    "                fullnbib = str()\n",
    "                for d in query:\n",
    "                    for i in range(1,pages):\n",
    "                        page = str(i)\n",
    "                        url = 'https://pubmed.ncbi.nlm.nih.gov/?term=' + d + '&format=pubmed&size=200&page='+ page\n",
    "                        output = rq.get(url)\n",
    "                        html = output.text\n",
    "                        soup = BeautifulSoup(html, features=\"html.parser\")\n",
    "                        for script in soup([\"script\", \"style\"]):\n",
    "                            script.extract()\n",
    "                        text = soup.get_text()\n",
    "                        postString = text.split(\"\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\",2)[2]\n",
    "                        nbib01 = postString.replace('\\n\\n','')\n",
    "                        fullnbib += nbib01\n",
    "                        request_counter += pages\n",
    "                        time.sleep(1.5)\n",
    "    \n",
    "                time2 = datetime.now()\n",
    "                timestamp = time2.strftime('%Y%m%d%H%M%S')\n",
    "                #^print('finish at:', datetime.now())\n",
    "                runtime = time2-time1\n",
    "                duration = str(runtime).split('.')[0]\n",
    "                hours, minutes, seconds = duration.split(':')\n",
    "                compact_duration = '{}:{}:{}'.format(hours, minutes, seconds)\n",
    "    \n",
    "    \n",
    "                # nbib parsing:\n",
    "                timea = datetime.now()\n",
    "                ref = nbib.read(fullnbib)\n",
    "                dfbib = pd.DataFrame(ref)\n",
    "                if 'descriptors' in dfbib.columns and len(dfbib['descriptors'])>1:\n",
    "                    dfbibdes = dfbib[['pubmed_id','descriptors']].dropna().reset_index(drop=True)\n",
    "                    nbib_objects = len(dfbib)\n",
    "                    nbib_objects_withdescriptors = len(dfbibdes)\n",
    "                    #print('nibib objects:',nbib_objects)\n",
    "                    #print('nibib objects with descriptors:',len(dfbibdes))\n",
    "                    timeb = datetime.now()\n",
    "                    #print('runtime:', timeb-timea)\n",
    "    \n",
    "                    #Statistics:\n",
    "                    pubmed_pmid_query = len(pmid_l)\n",
    "                    pubmed_pmid_nbib = len(dfbib.pubmed_id.drop_duplicates())\n",
    "                    pubmed_pmid_nbib_yesmesh = len(dfbibdes.pubmed_id.drop_duplicates())\n",
    "                    pubmed_pmid_nbib_nomesh = len(dfbib.pubmed_id.drop_duplicates())-len(dfbibdes.pubmed_id.drop_duplicates())\n",
    "    \n",
    "                    # refine MESH\n",
    "                    dfr = []\n",
    "                    for i in range(len(dfbibdes)):\n",
    "                        for mesh in dfbibdes['descriptors'][i]:\n",
    "                            out = dfbibdes['pubmed_id'][i], mesh\n",
    "                            dfr.append(out)\n",
    "                    MESH = pd.DataFrame(dfr).rename(columns={0: 'pmids',1:'mesh'})\n",
    "    \n",
    "                    # dataframe parsing splitting three fields\n",
    "                    MESHsplit =[]\n",
    "                    for i in range(len(MESH)):\n",
    "                        mg = MESH.mesh[i].get('descriptor')\n",
    "                        mg2 = MESH.mesh[i].get('qualifier')\n",
    "                        mg3 = MESH.mesh[i].get('major')\n",
    "                        mgg = MESH.pmids[i], mg, mg2, mg3\n",
    "                        MESHsplit.append(mgg)\n",
    "    \n",
    "                    dfmesh = pd.DataFrame(MESHsplit).rename(columns={0: 'pmids',1:'mesh',2:'qualifier',3:'major'}).drop_duplicates()\n",
    "    \n",
    "                    #statistics\n",
    "                    pubmed_pmidmesh = len(dfmesh[['pmids','mesh']].drop_duplicates())\n",
    "                    pubmed_mesh_qualifier_major = len(MESH.mesh.drop_duplicates())\n",
    "                    pubmed_mesh = len(dfmesh.mesh.drop_duplicates())\n",
    "    \n",
    "                    pmidmesh = dfmesh[['pmids','mesh']].drop_duplicates()\n",
    "                    pmidmesh['pmids'] = pmidmesh['pmids'].astype(str) #convert pmid type in str\n",
    "    \n",
    "    \n",
    "                    #Analyze enrichment with groupby.describe method-------------------------------\n",
    "                    #Add rsid coulmn con merge\n",
    "                    rspmidmesh_merge = pd.merge(pmidmesh, rsidpmid, on= 'pmids', how='inner').drop_duplicates().reindex(columns=['pmids', 'rsid', 'mesh'])\n",
    "                    #rspmidmesh_merge['pmids'] = rspmidmesh_merge['pmids'].astype(str)\n",
    "    \n",
    "                    ### groupby.describe analysis by mesh\n",
    "                    meshrspmidmerge_count = rspmidmesh_merge.groupby('mesh').describe().reset_index()\n",
    "                    meshrspmidmerge_count.columns = meshrspmidmerge_count.columns.to_flat_index()\n",
    "                    #to handle generate df.groupby.describe, convert Multicolumn to single column\n",
    "                    #https://datascientyst.com/flatten-multiindex-in-pandas/\n",
    "                    new_column_names = ['mesh', 'pmid-count', 'pmid-unique','pmid-top','pmid-freq','rsid-count', 'rsid-unique','rsid-top','rsid-freq']\n",
    "                    meshrspmidmerge_count.columns = new_column_names\n",
    "    \n",
    "                    meshrspmidmerge_count_short = meshrspmidmerge_count[['mesh','pmid-unique','rsid-unique']]\n",
    "                    #pmidmeshintmerge2meshlesssort = pmidmeshintmerge2meshless.sort_values(by='pmid-unique',ascending=False).reset_index(drop=True)\n",
    "    \n",
    "                    # add frequency\n",
    "                    totalpmid_count = len(pmidmesh.pmids.drop_duplicates())\n",
    "                    meshrspmidmerge_count_short_freq = meshrspmidmerge_count_short.copy()\n",
    "                    meshb_frq = meshrspmidmerge_count_short_freq.loc[:,'pmid-unique'].astype(float)/totalpmid_count\n",
    "                    meshrspmidmerge_count_short_freq.loc[:,'mesh frequency'] = round(meshb_frq,3)#*100\n",
    "                    meshrspmidmerge_count_short_freq_sort = meshrspmidmerge_count_short_freq.sort_values(by='pmid-unique',ascending=False).reset_index(drop=True)\n",
    "    \n",
    "                    top10mesh_all = meshrspmidmerge_count_short_freq_sort['mesh'][:10].tolist()\n",
    "                    #display(meshrspmidmerge_count_short_freq_sort.head(20))\n",
    "    \n",
    "                    ### groupby.describe analysis by rsid------------------\n",
    "                    rspmidmeshmerge_count = rspmidmesh_merge.groupby('rsid').describe().reset_index()\n",
    "                    rspmidmeshmerge_count.columns = rspmidmeshmerge_count.columns.to_flat_index()\n",
    "                    new_column_names = ['rsid', 'pmid-count', 'pmid-unique','pmid-top','pmid-freq','mesh-count', 'mesh-unique','mesh-top','mesh-freq']\n",
    "                    rspmidmeshmerge_count.columns = new_column_names\n",
    "    \n",
    "                    rsid_pmid10 = len(rspmidmeshmerge_count[rspmidmeshmerge_count['pmid-unique']>10])\n",
    "                    rsid_pmid50 = len(rspmidmeshmerge_count[rspmidmeshmerge_count['pmid-unique']>50])\n",
    "                    rsid_pmid100 = len(rspmidmeshmerge_count[rspmidmeshmerge_count['pmid-unique']>100])\n",
    "    \n",
    "                    rspmidmeshmerge_count_short = rspmidmeshmerge_count[['rsid','pmid-unique','mesh-unique']]\n",
    "                    rspmidmeshmerge_count_short_sort = rspmidmeshmerge_count_short.sort_values(by='pmid-unique', ascending= False).reset_index(drop=True)\n",
    "                    top10rsid_all = rspmidmeshmerge_count_short_sort['rsid'].iloc[:10].tolist()\n",
    "    \n",
    "                    if save_plot:\n",
    "                        # create a scatter plot-----------------------------------------\n",
    "                        x1 = meshrspmidmerge_count_short_freq_sort['mesh'].head(30)\n",
    "                        y1 = meshrspmidmerge_count_short_freq_sort['pmid-unique'].head(30)\n",
    "                        plt.figure(figsize=(5, 8))\n",
    "                        plt.title('Scatter Plot: '+gene+' pmid-mesh (total)', loc='center',pad=10)\n",
    "                        plt.scatter(y1, x1)\n",
    "                        plt.gca().invert_yaxis()\n",
    "                        #plt.yticks(rotation=90)\n",
    "                        plt.tick_params(axis='x', which='both', top=True, bottom=False, labeltop=True, labelbottom=False)\n",
    "                        plt.xlabel('pmid count', position=(0.5, 1.08))\n",
    "                        ax = plt.gca()\n",
    "                        ax.xaxis.set_label_position('top')\n",
    "                        plt.savefig(db_path+'/'+gene+'_mesh_plot_'+timestamp+'_total.png',dpi=120, bbox_inches = \"tight\")\n",
    "                        plt.close()\n",
    "    \n",
    "                    if save_studytype_data:\n",
    "                        # GET STUDY TYPE from NBIB----------------------------------------\n",
    "                        dfbib = pd.DataFrame(ref)\n",
    "                        dfbib.pubmed_id = dfbib.pubmed_id.astype('str')\n",
    "                        if 'publication_types' in dfbib.columns and len(dfbib['publication_types'])>1:\n",
    "                            dfbib_studyty = dfbib[['pubmed_id','publication_types']].dropna().reset_index(drop=True)\n",
    "    \n",
    "                            #PMID-Studytype table build:\n",
    "                            df_studytype = []\n",
    "                            for i in range(len(dfbib_studyty)):\n",
    "                                for studytype in dfbib_studyty['publication_types'][i]:\n",
    "                                    out = dfbib_studyty['pubmed_id'][i], studytype\n",
    "                                    df_studytype.append(out)\n",
    "                            STUDYT = pd.DataFrame(df_studytype).rename(columns={0: 'pmids',1:'study_type'})\n",
    "                            mask_st = STUDYT['study_type'].str.contains('Research Support|Journal Article')\n",
    "                            STUDYTless = STUDYT[~mask_st].reset_index(drop=True)\n",
    "    \n",
    "                            mask_lessing = STUDYT['pmids'].isin(STUDYTless['pmids'])\n",
    "                            STUDYTdiff = STUDYT[~mask_lessing].reset_index(drop=True)\n",
    "                            STUDYTdiff['study_type2'] = 'Unknown'\n",
    "                            STUDYTdiff = STUDYTdiff[['pmids','study_type2']].rename(columns={'study_type2':'study_type'}).drop_duplicates().reset_index(drop=True)\n",
    "                            #len(STUDYTless.pmids.drop_duplicates()), len(dfbib.pubmed_id.drop_duplicates())\n",
    "                            STUDYTconcat = pd.concat([STUDYTless, STUDYTdiff], ignore_index=True)\n",
    "                            STUDYTconcat#.pmids.drop_duplicates()\n",
    "    \n",
    "                            #study type count:\n",
    "                            STUDYTless_count = STUDYTconcat.groupby('study_type').describe().reset_index()\n",
    "                            STUDYTless_count.columns = STUDYTless_count.columns.to_flat_index()\n",
    "                            new_column_names = ['study_type', 'pmid-count', 'pmid-unique','pmid-top','pmid-freq']\n",
    "                            STUDYTless_count.columns = new_column_names\n",
    "                            STUDYTless_count_sort = STUDYTless_count.sort_values(by= 'pmid-count',ascending=False)\n",
    "    \n",
    "                            # save data\n",
    "                            STUDYTconcat.to_csv(db_path+'/'+gene+'_lit1pmid_studytype.csv')\n",
    "                        else:\n",
    "                            print(gene+' no publication_types in nbib')\n",
    "                            pass\n",
    "    \n",
    "    #SAVE TABLES-----------------------------------------------------------\n",
    "                    timestamp = time2.strftime('%Y%m%d%H%M%S')\n",
    "                    # save accessory data:\n",
    "                    if save_accessory_data:\n",
    "                        dfsort[[\"_id\",\"rsid\",\"pmids_count\"]].to_csv(db_path+'/'+gene+'_litvar2_variants4gene.csv')\n",
    "                        rsidpmid.to_csv(db_path+'/'+gene+'_litvar1_rsids2pmids.csv') #lit1 [rsid-pmid]\n",
    "                        #rsidpmidcountfsort #lit1 pmid count\n",
    "    \n",
    "                        meshrspmidmerge_count_short_freq_sort.to_csv(db_path+'/'+gene+'_mesh_pmidrsid_count.csv')\n",
    "    \n",
    "                    #complete_df with concat:\n",
    "                    #import gene-rsidpmidmesh and gene-rsidpmid\n",
    "                    dfmesh['pmids'] = dfmesh['pmids'].astype(str)\n",
    "                    rsidpmid['pmids'] = rsidpmid['pmids'].astype(str)\n",
    "    \n",
    "                    # add a rsid-merger to dfmesh\n",
    "                    gene_rsidpmidmesh = pd.merge(rsidpmid, dfmesh, on='pmids')\n",
    "                    gene_rsidpmidmesh['gene'] = gene\n",
    "    \n",
    "                    gene_df = pd.DataFrame(gene_rsidpmidmesh)\n",
    "                    complete_df = pd.concat([complete_df, gene_rsidpmidmesh])\n",
    "    \n",
    "                    #complete_nbibtable with concat:\n",
    "                    dfbib['gene'] = gene\n",
    "                    complete_nbibtable = pd.concat([complete_nbibtable, dfbib])\n",
    "                    #pyperclip.copy(str(dfbib.columns.to_list()))\n",
    "    \n",
    "                    # save checkpoint----------------------\n",
    "                    if genes.index(gene) > 1 and genes.index(gene) % checkpoint == 0:\n",
    "                        complete_df = complete_df.reindex(columns=['gene','rsid', 'pmids', 'mesh', 'qualifier', 'major'])\n",
    "                        complete_df.to_csv(db_path+'/grpm_table_output.csv')\n",
    "    \n",
    "                        complete_nbibtable = complete_nbibtable.reindex(columns=['gene','pubmed_id', 'citation_owner', 'nlm_status', 'last_revision_date', 'electronic_issn', 'linking_issn', 'journal_volume', 'journal_issue', 'publication_date', 'title', 'abstract', 'authors', 'language', 'grants', 'publication_types', 'electronic_publication_date', 'place_of_publication', 'journal_abbreviated', 'journal', 'nlm_journal_id', 'descriptors', 'pmcid', 'keywords', 'conflict_of_interest', 'received_time', 'revised_time', 'accepted_time', 'pubmed_time', 'medline_time', 'entrez_time', 'pii', 'doi', 'publication_status', 'print_issn', 'pages'])\n",
    "                        complete_nbibtable.to_csv(db_path+'/complete_nbibtable.csv')\n",
    "                        print(\"saved checkpoint\")\n",
    "                    else:\n",
    "                        pass\n",
    "    \n",
    "    \n",
    "    #REPORT-----------------------------------------------------------------\n",
    "                    time_omega = datetime.now()\n",
    "                    full_runtime = time_omega - time_alpha\n",
    "                    #^print('total runtime:', full_runtime)\n",
    "                    print(gene + '_runtime:', full_runtime)\n",
    "                    nbib_seconds = runtime.total_seconds()\n",
    "                    total_seconds = full_runtime.total_seconds()\n",
    "                    full_runtime_str = str(full_runtime).split('.')[0]\n",
    "    \n",
    "                    report = {'ncbi_dbsnp': NCBI_dbSNP,\n",
    "                              'lit2_variant': lit2_variant,\n",
    "                              'lit2_variant_norsid': lit2_variant_norsid,\n",
    "                              'lit2_rsid': lit2_rsid,\n",
    "                              'lit2_rsid_plus1': lit2_rsid_f,\n",
    "                              'lit1_rsid': lit1_rsid,\n",
    "                              #'lit1_raw_pmid': lit1_raw_pmid,\n",
    "                              #'lit1_rsid_pmid': lit1_rsid_pmid,\n",
    "                              'lit1_rsid_pmid_plus1': lit1_rsid_f,\n",
    "                              #lit1_rsid_pmid=1': lit1_rsid_m,\n",
    "                              'lit1_pmid': lit1_pmid,\n",
    "                              'lit1_pmid_pmid_plus1': lit1_pmid_f,\n",
    "                              'pubmed_pmid_query': pubmed_pmid_query,\n",
    "                              'nbib_objects': nbib_objects,\n",
    "                              'nbib_objects_withdescriptors': nbib_objects_withdescriptors,\n",
    "                              'pubmed_pmid': pubmed_pmid_nbib,\n",
    "                              'pubmed_pmid_withmesh': pubmed_pmid_nbib_yesmesh,\n",
    "                              #'pubmed_pmid_nomesh':pubmed_pmid_nbib_nomesh,\n",
    "                              'pubmed_pmidmesh': pubmed_pmidmesh,\n",
    "                              'pubmed_mesh_qualifier_major': pubmed_mesh_qualifier_major,\n",
    "                              'pubmed_mesh': pubmed_mesh,\n",
    "                              'rsid_pmid10': rsid_pmid10,\n",
    "                              'rsid_pmid50': rsid_pmid50,\n",
    "                              'rsid_pmid100': rsid_pmid100,\n",
    "                              'top10mesh_all': str(top10mesh_all),\n",
    "                              'top10rsid_all': str(top10rsid_all),\n",
    "                              'pubmed_runtime': duration,\n",
    "                              'total_runtime': full_runtime_str,\n",
    "                              'time_stamp': time2\n",
    "                              }\n",
    "    \n",
    "                    df_report = pd.DataFrame(report, index=[gene]).transpose()\n",
    "    \n",
    "                    # generate fist report.csv\n",
    "                    if os.path.isfile(db_path+'/GRPM_report.csv'):\n",
    "                        dfL = pd.read_csv(db_path+'/GRPM_report.csv', index_col=0)\n",
    "                        dfL = pd.concat([dfL, df_report], axis=1)\n",
    "                        dfL.to_csv(db_path+'/GRPM_report.csv')\n",
    "                    else:\n",
    "                        df_report.to_csv(db_path+'/GRPM_report.csv')  # solo la prima volta\n",
    "    \n",
    "                    #Update gene values\n",
    "                    GRPM_report = pd.read_csv(db_path+'/GRPM_report.csv', index_col=0)\n",
    "                    if gene + '.1' in GRPM_report.columns:\n",
    "                        GRPM_report = GRPM_report.drop(columns=gene)\n",
    "                        GRPM_report = GRPM_report.rename(columns={gene + '.1': gene})\n",
    "                        GRPM_report.to_csv(db_path+'/GRPM_report.csv')\n",
    "                        print(gene,'already in db')\n",
    "    \n",
    "                else:\n",
    "                    print(gene + ' no descriptors in nbib')\n",
    "                    time.sleep(0.8)\n",
    "                    pass\n",
    "            else:\n",
    "                print(gene + ' no results on litvar2 rsid2pmids')\n",
    "                time.sleep(0.8)\n",
    "                pass\n",
    "    else:\n",
    "        print(gene + ' no results on litvar2 gene2pmidcount')\n",
    "        pass\n",
    "\n",
    "    if request_counter > 9000:\n",
    "        dada = 2\n",
    "        #print('Allert! Reaching pubmed request limit')\n",
    "    if request_counter > 9950:\n",
    "        #print('Request limit reached. Wait \\'till tomorrow!')\n",
    "        time_finish = datetime.now()\n",
    "        time_batch = time_finish - time_start\n",
    "        time_batch_str = str(time_batch).split('.')[0]\n",
    "        #print('time batch:', time_batch_str)\n",
    "        #break\n",
    "\n",
    "complete_df = complete_df.reindex(columns=['gene','rsid', 'pmids', 'mesh', 'qualifier', 'major'])\n",
    "complete_df.to_csv(db_path+'/grpm_table_output.csv')\n",
    "\n",
    "complete_nbibtable = complete_nbibtable.reindex(columns=['gene','pubmed_id', 'citation_owner', 'nlm_status', 'last_revision_date', 'electronic_issn', 'linking_issn', 'journal_volume', 'journal_issue', 'publication_date', 'title', 'abstract', 'authors', 'language', 'grants', 'publication_types', 'electronic_publication_date', 'place_of_publication', 'journal_abbreviated', 'journal', 'nlm_journal_id', 'descriptors', 'pmcid', 'keywords', 'conflict_of_interest', 'received_time', 'revised_time', 'accepted_time', 'pubmed_time', 'medline_time', 'entrez_time', 'pii', 'doi', 'publication_status', 'print_issn', 'pages'])\n",
    "complete_nbibtable.to_csv(db_path+'/complete_nbibtable.csv')\n",
    "\n",
    "time_finish = datetime.now()\n",
    "time_batch = time_finish - time_start\n",
    "time_batch_str = str(time_batch).split('.')[0]\n",
    "print('gene batch:', len(genes))\n",
    "print('time batch:', time_batch_str)\n",
    "print('runtime/gene:', time_batch/len(genes))\n",
    "print('request_counter:', request_counter,' (limit: 10.000/day)')\n",
    "gene_counter += len(genes)\n",
    "print('requests/gene:', request_counter/gene_counter)\n",
    "print(time_finish)\n",
    "\n",
    "### notes:\n",
    "# LIMITS PubMed Programming Utilities (PMU)\n",
    "# 10 requests/second\n",
    "# 10,000 requests/day"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-03-14T11:19:52.878896200Z",
     "start_time": "2024-03-14T11:19:52.727955900Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "grpm report:\n"
     ]
    },
    {
     "data": {
      "text/plain": "         ncbi_dbsnp lit2_variant lit2_variant_norsid lit2_rsid  \\\nKIAA1143          0           54                   0        54   \nOPRL1             0          309                  56       251   \nATP23            na           92                   4        87   \nCENATAC          na          150                   0       150   \nZFY              na           61                  21        40   \nNLRP2            na          706                  22       661   \nISX              na          135                   0       133   \nANKRD12          na          594                   4       589   \nSFI1             na         1208                  34      1166   \nSTC1             na          179                  87        92   \nNEU1             na         4602                4220       297   \nCCL3L3           na           37                  19        18   \nGSTT4            na          126                   1       125   \nGSTT2            na           36                   1        35   \nZC3H3            na          643                   0       639   \nPRAMEF2          na          345                   0       338   \nADAM18           na          317                   0       316   \nZBED4            na          425                   1       424   \nPRPF31           na         1188                 315       831   \nMMEL1            na          456                  21       432   \nNPIPA7           na           24                   8        15   \nGGNBP2           na          169                   8       161   \n\n         lit2_rsid_plus1 lit1_rsid lit1_rsid_pmid_plus1 lit1_pmid  \\\nKIAA1143              27         0                    1         9   \nOPRL1                 98         0                    1         3   \nATP23                 59         0                    1         2   \nCENATAC               61         0                    1        13   \nZFY                   11         0                    1         2   \nNLRP2                307         0                    1         2   \nISX                   73         0                    1         2   \nANKRD12              179         0                    1         2   \nSFI1                 409         0                    1         2   \nSTC1                  32         0                    1         2   \nNEU1                 139         0                    1         2   \nCCL3L3                 6         0                    1         6   \nGSTT4                 62         0                    1         2   \nGSTT2                 23         0                    1         7   \nZC3H3                317         0                    1         5   \nPRAMEF2              174         0                    1         2   \nADAM18               123         0                    1         3   \nZBED4                212         0                    1         3   \nPRPF31               406         0                    1         4   \nMMEL1                213         0                    1         4   \nNPIPA7                 3         0                    1         2   \nGGNBP2                69         0                    1        30   \n\n         lit1_pmid_pmid_plus1 pubmed_pmid_query  ...  \\\nKIAA1143                    9                 9  ...   \nOPRL1                       3                 3  ...   \nATP23                       2                 2  ...   \nCENATAC                    13                13  ...   \nZFY                         2                 2  ...   \nNLRP2                       2                 2  ...   \nISX                         2                 2  ...   \nANKRD12                     2                 2  ...   \nSFI1                        2                 2  ...   \nSTC1                        2                 2  ...   \nNEU1                        2                 2  ...   \nCCL3L3                      6                 6  ...   \nGSTT4                       2                 2  ...   \nGSTT2                       7                 7  ...   \nZC3H3                       5                 5  ...   \nPRAMEF2                     2                 2  ...   \nADAM18                      3                 3  ...   \nZBED4                       3                 3  ...   \nPRPF31                      4                 4  ...   \nMMEL1                       4                 4  ...   \nNPIPA7                      2                 2  ...   \nGGNBP2                     30                30  ...   \n\n         pubmed_mesh_qualifier_major pubmed_mesh rsid_pmid10 rsid_pmid50  \\\nKIAA1143                          89          72           0           0   \nOPRL1                             77          65           0           0   \nATP23                             36          30           0           0   \nCENATAC                           87          69           1           0   \nZFY                               10          10           0           0   \nNLRP2                             16          13           0           0   \nISX                               32          29           0           0   \nANKRD12                           25          23           0           0   \nSFI1                              17          12           0           0   \nSTC1                              24          24           0           0   \nNEU1                              34          25           0           0   \nCCL3L3                            46          35           0           0   \nGSTT4                             26          26           0           0   \nGSTT2                             68          55           0           0   \nZC3H3                             70          51           0           0   \nPRAMEF2                           14          14           0           0   \nADAM18                            60          45           0           0   \nZBED4                             25          25           0           0   \nPRPF31                            85          59           0           0   \nMMEL1                             43          36           0           0   \nNPIPA7                            22          18           0           0   \nGGNBP2                           272         186           1           0   \n\n         rsid_pmid100                                      top10mesh_all  \\\nKIAA1143            0  ['Humans', 'Quantitative Trait Loci', 'Polymor...   \nOPRL1               0  ['Signal Transduction', 'Mutation', 'Humans', ...   \nATP23               0  ['Mutation', 'Humans', 'Adult', 'Animals', 'Up...   \nCENATAC             0  ['Humans', 'Genetic Predisposition to Disease'...   \nZFY                 0  ['Chromosomes, Human, Y', 'Databases, Genetic'...   \nNLRP2               0  ['Animals', 'Cell Line, Tumor', 'Fatty Acid De...   \nISX                 0  ['Humans', 'Adenocarcinoma', 'High-Throughput ...   \nANKRD12             0  ['Humans', 'Image Processing, Computer-Assiste...   \nSFI1                0  ['Animals', 'CD8-Positive T-Lymphocytes', 'Cel...   \nSTC1                0  ['Polymorphism, Single Nucleotide', 'Humans', ...   \nNEU1                0  ['HLA-DRB1 Chains', 'Polymorphism, Single Nucl...   \nCCL3L3              0  ['Polymorphism, Single Nucleotide', 'Oryza', '...   \nGSTT4               0  ['Humans', 'Adolescent', 'Adult', 'Segmental D...   \nGSTT2               0  ['Humans', 'Exome', 'High-Throughput Nucleotid...   \nZC3H3               0  ['Humans', 'Quantitative Trait Loci', 'White P...   \nPRAMEF2             0  ['Humans', 'Aged', 'Aged, 80 and over', 'Bisph...   \nADAM18              0  ['Humans', 'Genome-Wide Association Study', 'G...   \nZBED4               0  ['Exome', 'Humans', 'Middle Aged', 'Aged', 'Hi...   \nPRPF31              0  ['Gene Expression Regulation, Neoplastic', 'Hu...   \nMMEL1               0  ['Humans', 'Ovarian Neoplasms', 'Computational...   \nNPIPA7              0  ['RNA Editing', 'Humans', 'APOBEC-3G Deaminase...   \nGGNBP2              0  ['Humans', 'Genome-Wide Association Study', 'P...   \n\n            top10rsid_all pubmed_runtime total_runtime  \\\nKIAA1143     ['rs937834']        0:00:02       0:00:05   \nOPRL1     ['rs949887692']        0:00:02       0:00:05   \nATP23     ['rs867626853']        0:00:02       0:00:04   \nCENATAC     ['rs9645664']        0:00:02       0:00:04   \nZFY       ['rs996955491']        0:00:02       0:00:04   \nNLRP2     ['rs996672212']        0:00:02       0:00:05   \nISX         ['rs9622120']        0:00:02       0:00:04   \nANKRD12      ['rs998172']        0:00:02       0:00:04   \nSFI1      ['rs979361662']        0:00:02       0:00:06   \nSTC1        ['rs9969426']        0:00:02       0:00:04   \nNEU1        ['rs9461727']        0:00:02       0:00:10   \nCCL3L3         ['rs2944']        0:00:02       0:00:04   \nGSTT4       ['rs9612533']        0:00:02       0:00:04   \nGSTT2      ['rs60718122']        0:00:02       0:00:04   \nZC3H3       ['rs9886386']        0:00:02       0:00:05   \nPRAMEF2     ['rs9729325']        0:00:02       0:00:04   \nADAM18      ['rs9657175']        0:00:02       0:00:04   \nZBED4       ['rs9627781']        0:00:02       0:00:05   \nPRPF31    ['rs990859359']        0:00:02       0:00:06   \nMMEL1     ['rs918533717']        0:00:02       0:00:05   \nNPIPA7    ['rs921840471']        0:00:02       0:00:04   \nGGNBP2      ['rs9906189']        0:00:02       0:00:05   \n\n                          time_stamp  \nKIAA1143  2024-03-14 12:12:48.587962  \nOPRL1     2024-03-14 12:12:55.485997  \nATP23     2024-03-14 12:18:07.128893  \nCENATAC   2024-03-14 12:18:12.096308  \nZFY       2024-03-14 12:18:16.705002  \nNLRP2     2024-03-14 12:18:22.263277  \nISX       2024-03-14 12:18:26.897125  \nANKRD12   2024-03-14 12:18:31.918605  \nSFI1      2024-03-14 12:18:38.396394  \nSTC1      2024-03-14 12:18:43.130012  \nNEU1      2024-03-14 12:18:53.603186  \nCCL3L3    2024-03-14 12:18:58.402481  \nGSTT4     2024-03-14 12:19:03.144046  \nGSTT2     2024-03-14 12:19:07.626489  \nZC3H3     2024-03-14 12:19:13.210048  \nPRAMEF2   2024-03-14 12:19:18.065832  \nADAM18    2024-03-14 12:19:22.852187  \nZBED4     2024-03-14 12:19:28.130703  \nPRPF31    2024-03-14 12:19:34.271767  \nMMEL1     2024-03-14 12:19:39.909325  \nNPIPA7    2024-03-14 12:19:44.243473  \nGGNBP2    2024-03-14 12:19:49.514856  \n\n[22 rows x 25 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>ncbi_dbsnp</th>\n      <th>lit2_variant</th>\n      <th>lit2_variant_norsid</th>\n      <th>lit2_rsid</th>\n      <th>lit2_rsid_plus1</th>\n      <th>lit1_rsid</th>\n      <th>lit1_rsid_pmid_plus1</th>\n      <th>lit1_pmid</th>\n      <th>lit1_pmid_pmid_plus1</th>\n      <th>pubmed_pmid_query</th>\n      <th>...</th>\n      <th>pubmed_mesh_qualifier_major</th>\n      <th>pubmed_mesh</th>\n      <th>rsid_pmid10</th>\n      <th>rsid_pmid50</th>\n      <th>rsid_pmid100</th>\n      <th>top10mesh_all</th>\n      <th>top10rsid_all</th>\n      <th>pubmed_runtime</th>\n      <th>total_runtime</th>\n      <th>time_stamp</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>KIAA1143</th>\n      <td>0</td>\n      <td>54</td>\n      <td>0</td>\n      <td>54</td>\n      <td>27</td>\n      <td>0</td>\n      <td>1</td>\n      <td>9</td>\n      <td>9</td>\n      <td>9</td>\n      <td>...</td>\n      <td>89</td>\n      <td>72</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Humans', 'Quantitative Trait Loci', 'Polymor...</td>\n      <td>['rs937834']</td>\n      <td>0:00:02</td>\n      <td>0:00:05</td>\n      <td>2024-03-14 12:12:48.587962</td>\n    </tr>\n    <tr>\n      <th>OPRL1</th>\n      <td>0</td>\n      <td>309</td>\n      <td>56</td>\n      <td>251</td>\n      <td>98</td>\n      <td>0</td>\n      <td>1</td>\n      <td>3</td>\n      <td>3</td>\n      <td>3</td>\n      <td>...</td>\n      <td>77</td>\n      <td>65</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Signal Transduction', 'Mutation', 'Humans', ...</td>\n      <td>['rs949887692']</td>\n      <td>0:00:02</td>\n      <td>0:00:05</td>\n      <td>2024-03-14 12:12:55.485997</td>\n    </tr>\n    <tr>\n      <th>ATP23</th>\n      <td>na</td>\n      <td>92</td>\n      <td>4</td>\n      <td>87</td>\n      <td>59</td>\n      <td>0</td>\n      <td>1</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>...</td>\n      <td>36</td>\n      <td>30</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Mutation', 'Humans', 'Adult', 'Animals', 'Up...</td>\n      <td>['rs867626853']</td>\n      <td>0:00:02</td>\n      <td>0:00:04</td>\n      <td>2024-03-14 12:18:07.128893</td>\n    </tr>\n    <tr>\n      <th>CENATAC</th>\n      <td>na</td>\n      <td>150</td>\n      <td>0</td>\n      <td>150</td>\n      <td>61</td>\n      <td>0</td>\n      <td>1</td>\n      <td>13</td>\n      <td>13</td>\n      <td>13</td>\n      <td>...</td>\n      <td>87</td>\n      <td>69</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Humans', 'Genetic Predisposition to Disease'...</td>\n      <td>['rs9645664']</td>\n      <td>0:00:02</td>\n      <td>0:00:04</td>\n      <td>2024-03-14 12:18:12.096308</td>\n    </tr>\n    <tr>\n      <th>ZFY</th>\n      <td>na</td>\n      <td>61</td>\n      <td>21</td>\n      <td>40</td>\n      <td>11</td>\n      <td>0</td>\n      <td>1</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>...</td>\n      <td>10</td>\n      <td>10</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Chromosomes, Human, Y', 'Databases, Genetic'...</td>\n      <td>['rs996955491']</td>\n      <td>0:00:02</td>\n      <td>0:00:04</td>\n      <td>2024-03-14 12:18:16.705002</td>\n    </tr>\n    <tr>\n      <th>NLRP2</th>\n      <td>na</td>\n      <td>706</td>\n      <td>22</td>\n      <td>661</td>\n      <td>307</td>\n      <td>0</td>\n      <td>1</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>...</td>\n      <td>16</td>\n      <td>13</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Animals', 'Cell Line, Tumor', 'Fatty Acid De...</td>\n      <td>['rs996672212']</td>\n      <td>0:00:02</td>\n      <td>0:00:05</td>\n      <td>2024-03-14 12:18:22.263277</td>\n    </tr>\n    <tr>\n      <th>ISX</th>\n      <td>na</td>\n      <td>135</td>\n      <td>0</td>\n      <td>133</td>\n      <td>73</td>\n      <td>0</td>\n      <td>1</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>...</td>\n      <td>32</td>\n      <td>29</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Humans', 'Adenocarcinoma', 'High-Throughput ...</td>\n      <td>['rs9622120']</td>\n      <td>0:00:02</td>\n      <td>0:00:04</td>\n      <td>2024-03-14 12:18:26.897125</td>\n    </tr>\n    <tr>\n      <th>ANKRD12</th>\n      <td>na</td>\n      <td>594</td>\n      <td>4</td>\n      <td>589</td>\n      <td>179</td>\n      <td>0</td>\n      <td>1</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>...</td>\n      <td>25</td>\n      <td>23</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Humans', 'Image Processing, Computer-Assiste...</td>\n      <td>['rs998172']</td>\n      <td>0:00:02</td>\n      <td>0:00:04</td>\n      <td>2024-03-14 12:18:31.918605</td>\n    </tr>\n    <tr>\n      <th>SFI1</th>\n      <td>na</td>\n      <td>1208</td>\n      <td>34</td>\n      <td>1166</td>\n      <td>409</td>\n      <td>0</td>\n      <td>1</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>...</td>\n      <td>17</td>\n      <td>12</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Animals', 'CD8-Positive T-Lymphocytes', 'Cel...</td>\n      <td>['rs979361662']</td>\n      <td>0:00:02</td>\n      <td>0:00:06</td>\n      <td>2024-03-14 12:18:38.396394</td>\n    </tr>\n    <tr>\n      <th>STC1</th>\n      <td>na</td>\n      <td>179</td>\n      <td>87</td>\n      <td>92</td>\n      <td>32</td>\n      <td>0</td>\n      <td>1</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>...</td>\n      <td>24</td>\n      <td>24</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Polymorphism, Single Nucleotide', 'Humans', ...</td>\n      <td>['rs9969426']</td>\n      <td>0:00:02</td>\n      <td>0:00:04</td>\n      <td>2024-03-14 12:18:43.130012</td>\n    </tr>\n    <tr>\n      <th>NEU1</th>\n      <td>na</td>\n      <td>4602</td>\n      <td>4220</td>\n      <td>297</td>\n      <td>139</td>\n      <td>0</td>\n      <td>1</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>...</td>\n      <td>34</td>\n      <td>25</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['HLA-DRB1 Chains', 'Polymorphism, Single Nucl...</td>\n      <td>['rs9461727']</td>\n      <td>0:00:02</td>\n      <td>0:00:10</td>\n      <td>2024-03-14 12:18:53.603186</td>\n    </tr>\n    <tr>\n      <th>CCL3L3</th>\n      <td>na</td>\n      <td>37</td>\n      <td>19</td>\n      <td>18</td>\n      <td>6</td>\n      <td>0</td>\n      <td>1</td>\n      <td>6</td>\n      <td>6</td>\n      <td>6</td>\n      <td>...</td>\n      <td>46</td>\n      <td>35</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Polymorphism, Single Nucleotide', 'Oryza', '...</td>\n      <td>['rs2944']</td>\n      <td>0:00:02</td>\n      <td>0:00:04</td>\n      <td>2024-03-14 12:18:58.402481</td>\n    </tr>\n    <tr>\n      <th>GSTT4</th>\n      <td>na</td>\n      <td>126</td>\n      <td>1</td>\n      <td>125</td>\n      <td>62</td>\n      <td>0</td>\n      <td>1</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>...</td>\n      <td>26</td>\n      <td>26</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Humans', 'Adolescent', 'Adult', 'Segmental D...</td>\n      <td>['rs9612533']</td>\n      <td>0:00:02</td>\n      <td>0:00:04</td>\n      <td>2024-03-14 12:19:03.144046</td>\n    </tr>\n    <tr>\n      <th>GSTT2</th>\n      <td>na</td>\n      <td>36</td>\n      <td>1</td>\n      <td>35</td>\n      <td>23</td>\n      <td>0</td>\n      <td>1</td>\n      <td>7</td>\n      <td>7</td>\n      <td>7</td>\n      <td>...</td>\n      <td>68</td>\n      <td>55</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Humans', 'Exome', 'High-Throughput Nucleotid...</td>\n      <td>['rs60718122']</td>\n      <td>0:00:02</td>\n      <td>0:00:04</td>\n      <td>2024-03-14 12:19:07.626489</td>\n    </tr>\n    <tr>\n      <th>ZC3H3</th>\n      <td>na</td>\n      <td>643</td>\n      <td>0</td>\n      <td>639</td>\n      <td>317</td>\n      <td>0</td>\n      <td>1</td>\n      <td>5</td>\n      <td>5</td>\n      <td>5</td>\n      <td>...</td>\n      <td>70</td>\n      <td>51</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Humans', 'Quantitative Trait Loci', 'White P...</td>\n      <td>['rs9886386']</td>\n      <td>0:00:02</td>\n      <td>0:00:05</td>\n      <td>2024-03-14 12:19:13.210048</td>\n    </tr>\n    <tr>\n      <th>PRAMEF2</th>\n      <td>na</td>\n      <td>345</td>\n      <td>0</td>\n      <td>338</td>\n      <td>174</td>\n      <td>0</td>\n      <td>1</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>...</td>\n      <td>14</td>\n      <td>14</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Humans', 'Aged', 'Aged, 80 and over', 'Bisph...</td>\n      <td>['rs9729325']</td>\n      <td>0:00:02</td>\n      <td>0:00:04</td>\n      <td>2024-03-14 12:19:18.065832</td>\n    </tr>\n    <tr>\n      <th>ADAM18</th>\n      <td>na</td>\n      <td>317</td>\n      <td>0</td>\n      <td>316</td>\n      <td>123</td>\n      <td>0</td>\n      <td>1</td>\n      <td>3</td>\n      <td>3</td>\n      <td>3</td>\n      <td>...</td>\n      <td>60</td>\n      <td>45</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Humans', 'Genome-Wide Association Study', 'G...</td>\n      <td>['rs9657175']</td>\n      <td>0:00:02</td>\n      <td>0:00:04</td>\n      <td>2024-03-14 12:19:22.852187</td>\n    </tr>\n    <tr>\n      <th>ZBED4</th>\n      <td>na</td>\n      <td>425</td>\n      <td>1</td>\n      <td>424</td>\n      <td>212</td>\n      <td>0</td>\n      <td>1</td>\n      <td>3</td>\n      <td>3</td>\n      <td>3</td>\n      <td>...</td>\n      <td>25</td>\n      <td>25</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Exome', 'Humans', 'Middle Aged', 'Aged', 'Hi...</td>\n      <td>['rs9627781']</td>\n      <td>0:00:02</td>\n      <td>0:00:05</td>\n      <td>2024-03-14 12:19:28.130703</td>\n    </tr>\n    <tr>\n      <th>PRPF31</th>\n      <td>na</td>\n      <td>1188</td>\n      <td>315</td>\n      <td>831</td>\n      <td>406</td>\n      <td>0</td>\n      <td>1</td>\n      <td>4</td>\n      <td>4</td>\n      <td>4</td>\n      <td>...</td>\n      <td>85</td>\n      <td>59</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Gene Expression Regulation, Neoplastic', 'Hu...</td>\n      <td>['rs990859359']</td>\n      <td>0:00:02</td>\n      <td>0:00:06</td>\n      <td>2024-03-14 12:19:34.271767</td>\n    </tr>\n    <tr>\n      <th>MMEL1</th>\n      <td>na</td>\n      <td>456</td>\n      <td>21</td>\n      <td>432</td>\n      <td>213</td>\n      <td>0</td>\n      <td>1</td>\n      <td>4</td>\n      <td>4</td>\n      <td>4</td>\n      <td>...</td>\n      <td>43</td>\n      <td>36</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Humans', 'Ovarian Neoplasms', 'Computational...</td>\n      <td>['rs918533717']</td>\n      <td>0:00:02</td>\n      <td>0:00:05</td>\n      <td>2024-03-14 12:19:39.909325</td>\n    </tr>\n    <tr>\n      <th>NPIPA7</th>\n      <td>na</td>\n      <td>24</td>\n      <td>8</td>\n      <td>15</td>\n      <td>3</td>\n      <td>0</td>\n      <td>1</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>...</td>\n      <td>22</td>\n      <td>18</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['RNA Editing', 'Humans', 'APOBEC-3G Deaminase...</td>\n      <td>['rs921840471']</td>\n      <td>0:00:02</td>\n      <td>0:00:04</td>\n      <td>2024-03-14 12:19:44.243473</td>\n    </tr>\n    <tr>\n      <th>GGNBP2</th>\n      <td>na</td>\n      <td>169</td>\n      <td>8</td>\n      <td>161</td>\n      <td>69</td>\n      <td>0</td>\n      <td>1</td>\n      <td>30</td>\n      <td>30</td>\n      <td>30</td>\n      <td>...</td>\n      <td>272</td>\n      <td>186</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>['Humans', 'Genome-Wide Association Study', 'P...</td>\n      <td>['rs9906189']</td>\n      <td>0:00:02</td>\n      <td>0:00:05</td>\n      <td>2024-03-14 12:19:49.514856</td>\n    </tr>\n  </tbody>\n</table>\n<p>22 rows  25 columns</p>\n</div>"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "grpm table genes: 22\n",
      "\n",
      "nbib table:\n"
     ]
    },
    {
     "data": {
      "text/plain": "        gene  pubmed_id citation_owner          nlm_status last_revision_date  \\\n0   KIAA1143   38225249            NLM             MEDLINE         2024-02-07   \n1   KIAA1143   30127527            NLM             MEDLINE         2020-03-09   \n2   KIAA1143   38190101            NLM             MEDLINE         2024-02-10   \n3   KIAA1143   25326100            NLM             MEDLINE         2022-12-07   \n4   KIAA1143   31127932            NLM             MEDLINE         2020-03-09   \n..       ...        ...            ...                 ...                ...   \n25    GGNBP2   24956270            NLM             MEDLINE         2022-12-07   \n26    GGNBP2   19128478            NLM  PubMed-not-MEDLINE         2021-10-20   \n27    GGNBP2   21829388            NLM             MEDLINE         2022-04-08   \n28    GGNBP2   25101798            NLM             MEDLINE         2023-12-13   \n29    GGNBP2   26019233            NLM             MEDLINE         2022-04-08   \n\n   electronic_issn linking_issn  journal_volume journal_issue  \\\n0        2041-1723    2041-1723              15             1   \n1        1546-1718    1061-4036              50             9   \n2        2666-979X    2666-979X               4             1   \n3        1744-4292    1744-4292              10            10   \n4        1460-2083    0964-6906              28            18   \n..             ...          ...             ...           ...   \n25       1932-6203    1932-6203               9             6   \n26       1755-8794    1755-8794               2           NaN   \n27       1553-7404    1553-7390               7             8   \n28       1476-5470    1466-4879              15             8   \n29       1460-2083    0964-6906              24            16   \n\n   publication_date  ... revised_time accepted_time          pubmed_time  \\\n0       2024 Jan 15  ...          NaN    2023-12-21  2024-01-16 00:42:00   \n1          2018 Sep  ...          NaN    2018-07-05  2018-08-22 06:00:00   \n2       2024 Jan 10  ...   2023-09-07    2023-11-19  2024-01-08 12:43:00   \n3       2014 Oct 16  ...          NaN           NaN  2014-10-19 06:00:00   \n4       2019 Sep 15  ...   2019-05-10    2019-05-20  2019-05-28 06:00:00   \n..              ...  ...          ...           ...                  ...   \n25             2014  ...          NaN    2014-06-02  2014-06-24 06:00:00   \n26       2009 Jan 7  ...          NaN    2009-01-07  2009-01-09 09:00:00   \n27         2011 Aug  ...          NaN    2011-06-06  2011-08-11 06:00:00   \n28         2014 Dec  ...   2014-06-24    2014-06-25  2014-08-08 06:00:00   \n29      2015 Aug 15  ...          NaN    2015-05-21  2015-05-29 06:00:00   \n\n           medline_time          entrez_time                        pii  \\\n0   2024-01-17 06:42:00  2024-01-15 23:15:00                      44680   \n1   2019-04-25 06:00:00  2018-08-22 06:00:00  10.1038/s41588-018-0192-y   \n2   2024-02-09 06:43:00  2024-01-08 11:27:00                     100465   \n3   2015-02-11 06:00:00  2014-10-19 06:00:00                        NaN   \n4   2020-03-10 06:00:00  2019-05-26 06:00:00                     ddz113   \n..                  ...                  ...                        ...   \n25  2015-10-13 06:00:00  2014-06-24 06:00:00            PONE-D-14-03246   \n26  2009-01-09 09:01:00  2009-01-09 09:00:00              1755-8794-2-1   \n27  2011-12-14 06:00:00  2011-08-11 06:00:00       PGENETICS-D-11-00416   \n28  2015-07-25 06:00:00  2014-08-08 06:00:00                 gene201447   \n29  2016-04-07 06:00:00  2015-05-29 06:00:00                     ddv194   \n\n                             doi publication_status print_issn      pages  \n0     10.1038/s41467-023-44680-3           epublish        NaN        528  \n1      10.1038/s41588-018-0192-y           ppublish  1061-4036  1327-1334  \n2     10.1016/j.xgen.2023.100465           ppublish        NaN     100465  \n3          10.15252/msb.20145114           epublish  1744-4292        754  \n4             10.1093/hmg/ddz113           ppublish  0964-6906  3037-3042  \n..                           ...                ...        ...        ...  \n25  10.1371/journal.pone.0100924           epublish        NaN    e100924  \n26         10.1186/1755-8794-2-1           epublish        NaN          1  \n27  10.1371/journal.pgen.1002197           ppublish  1553-7390   e1002197  \n28          10.1038/gene.2014.47           ppublish  1466-4879     534-42  \n29            10.1093/hmg/ddv194           ppublish  0964-6906    4746-63  \n\n[109 rows x 36 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>gene</th>\n      <th>pubmed_id</th>\n      <th>citation_owner</th>\n      <th>nlm_status</th>\n      <th>last_revision_date</th>\n      <th>electronic_issn</th>\n      <th>linking_issn</th>\n      <th>journal_volume</th>\n      <th>journal_issue</th>\n      <th>publication_date</th>\n      <th>...</th>\n      <th>revised_time</th>\n      <th>accepted_time</th>\n      <th>pubmed_time</th>\n      <th>medline_time</th>\n      <th>entrez_time</th>\n      <th>pii</th>\n      <th>doi</th>\n      <th>publication_status</th>\n      <th>print_issn</th>\n      <th>pages</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>KIAA1143</td>\n      <td>38225249</td>\n      <td>NLM</td>\n      <td>MEDLINE</td>\n      <td>2024-02-07</td>\n      <td>2041-1723</td>\n      <td>2041-1723</td>\n      <td>15</td>\n      <td>1</td>\n      <td>2024 Jan 15</td>\n      <td>...</td>\n      <td>NaN</td>\n      <td>2023-12-21</td>\n      <td>2024-01-16 00:42:00</td>\n      <td>2024-01-17 06:42:00</td>\n      <td>2024-01-15 23:15:00</td>\n      <td>44680</td>\n      <td>10.1038/s41467-023-44680-3</td>\n      <td>epublish</td>\n      <td>NaN</td>\n      <td>528</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>KIAA1143</td>\n      <td>30127527</td>\n      <td>NLM</td>\n      <td>MEDLINE</td>\n      <td>2020-03-09</td>\n      <td>1546-1718</td>\n      <td>1061-4036</td>\n      <td>50</td>\n      <td>9</td>\n      <td>2018 Sep</td>\n      <td>...</td>\n      <td>NaN</td>\n      <td>2018-07-05</td>\n      <td>2018-08-22 06:00:00</td>\n      <td>2019-04-25 06:00:00</td>\n      <td>2018-08-22 06:00:00</td>\n      <td>10.1038/s41588-018-0192-y</td>\n      <td>10.1038/s41588-018-0192-y</td>\n      <td>ppublish</td>\n      <td>1061-4036</td>\n      <td>1327-1334</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>KIAA1143</td>\n      <td>38190101</td>\n      <td>NLM</td>\n      <td>MEDLINE</td>\n      <td>2024-02-10</td>\n      <td>2666-979X</td>\n      <td>2666-979X</td>\n      <td>4</td>\n      <td>1</td>\n      <td>2024 Jan 10</td>\n      <td>...</td>\n      <td>2023-09-07</td>\n      <td>2023-11-19</td>\n      <td>2024-01-08 12:43:00</td>\n      <td>2024-02-09 06:43:00</td>\n      <td>2024-01-08 11:27:00</td>\n      <td>100465</td>\n      <td>10.1016/j.xgen.2023.100465</td>\n      <td>ppublish</td>\n      <td>NaN</td>\n      <td>100465</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>KIAA1143</td>\n      <td>25326100</td>\n      <td>NLM</td>\n      <td>MEDLINE</td>\n      <td>2022-12-07</td>\n      <td>1744-4292</td>\n      <td>1744-4292</td>\n      <td>10</td>\n      <td>10</td>\n      <td>2014 Oct 16</td>\n      <td>...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>2014-10-19 06:00:00</td>\n      <td>2015-02-11 06:00:00</td>\n      <td>2014-10-19 06:00:00</td>\n      <td>NaN</td>\n      <td>10.15252/msb.20145114</td>\n      <td>epublish</td>\n      <td>1744-4292</td>\n      <td>754</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>KIAA1143</td>\n      <td>31127932</td>\n      <td>NLM</td>\n      <td>MEDLINE</td>\n      <td>2020-03-09</td>\n      <td>1460-2083</td>\n      <td>0964-6906</td>\n      <td>28</td>\n      <td>18</td>\n      <td>2019 Sep 15</td>\n      <td>...</td>\n      <td>2019-05-10</td>\n      <td>2019-05-20</td>\n      <td>2019-05-28 06:00:00</td>\n      <td>2020-03-10 06:00:00</td>\n      <td>2019-05-26 06:00:00</td>\n      <td>ddz113</td>\n      <td>10.1093/hmg/ddz113</td>\n      <td>ppublish</td>\n      <td>0964-6906</td>\n      <td>3037-3042</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>25</th>\n      <td>GGNBP2</td>\n      <td>24956270</td>\n      <td>NLM</td>\n      <td>MEDLINE</td>\n      <td>2022-12-07</td>\n      <td>1932-6203</td>\n      <td>1932-6203</td>\n      <td>9</td>\n      <td>6</td>\n      <td>2014</td>\n      <td>...</td>\n      <td>NaN</td>\n      <td>2014-06-02</td>\n      <td>2014-06-24 06:00:00</td>\n      <td>2015-10-13 06:00:00</td>\n      <td>2014-06-24 06:00:00</td>\n      <td>PONE-D-14-03246</td>\n      <td>10.1371/journal.pone.0100924</td>\n      <td>epublish</td>\n      <td>NaN</td>\n      <td>e100924</td>\n    </tr>\n    <tr>\n      <th>26</th>\n      <td>GGNBP2</td>\n      <td>19128478</td>\n      <td>NLM</td>\n      <td>PubMed-not-MEDLINE</td>\n      <td>2021-10-20</td>\n      <td>1755-8794</td>\n      <td>1755-8794</td>\n      <td>2</td>\n      <td>NaN</td>\n      <td>2009 Jan 7</td>\n      <td>...</td>\n      <td>NaN</td>\n      <td>2009-01-07</td>\n      <td>2009-01-09 09:00:00</td>\n      <td>2009-01-09 09:01:00</td>\n      <td>2009-01-09 09:00:00</td>\n      <td>1755-8794-2-1</td>\n      <td>10.1186/1755-8794-2-1</td>\n      <td>epublish</td>\n      <td>NaN</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>27</th>\n      <td>GGNBP2</td>\n      <td>21829388</td>\n      <td>NLM</td>\n      <td>MEDLINE</td>\n      <td>2022-04-08</td>\n      <td>1553-7404</td>\n      <td>1553-7390</td>\n      <td>7</td>\n      <td>8</td>\n      <td>2011 Aug</td>\n      <td>...</td>\n      <td>NaN</td>\n      <td>2011-06-06</td>\n      <td>2011-08-11 06:00:00</td>\n      <td>2011-12-14 06:00:00</td>\n      <td>2011-08-11 06:00:00</td>\n      <td>PGENETICS-D-11-00416</td>\n      <td>10.1371/journal.pgen.1002197</td>\n      <td>ppublish</td>\n      <td>1553-7390</td>\n      <td>e1002197</td>\n    </tr>\n    <tr>\n      <th>28</th>\n      <td>GGNBP2</td>\n      <td>25101798</td>\n      <td>NLM</td>\n      <td>MEDLINE</td>\n      <td>2023-12-13</td>\n      <td>1476-5470</td>\n      <td>1466-4879</td>\n      <td>15</td>\n      <td>8</td>\n      <td>2014 Dec</td>\n      <td>...</td>\n      <td>2014-06-24</td>\n      <td>2014-06-25</td>\n      <td>2014-08-08 06:00:00</td>\n      <td>2015-07-25 06:00:00</td>\n      <td>2014-08-08 06:00:00</td>\n      <td>gene201447</td>\n      <td>10.1038/gene.2014.47</td>\n      <td>ppublish</td>\n      <td>1466-4879</td>\n      <td>534-42</td>\n    </tr>\n    <tr>\n      <th>29</th>\n      <td>GGNBP2</td>\n      <td>26019233</td>\n      <td>NLM</td>\n      <td>MEDLINE</td>\n      <td>2022-04-08</td>\n      <td>1460-2083</td>\n      <td>0964-6906</td>\n      <td>24</td>\n      <td>16</td>\n      <td>2015 Aug 15</td>\n      <td>...</td>\n      <td>NaN</td>\n      <td>2015-05-21</td>\n      <td>2015-05-29 06:00:00</td>\n      <td>2016-04-07 06:00:00</td>\n      <td>2015-05-29 06:00:00</td>\n      <td>ddv194</td>\n      <td>10.1093/hmg/ddv194</td>\n      <td>ppublish</td>\n      <td>0964-6906</td>\n      <td>4746-63</td>\n    </tr>\n  </tbody>\n</table>\n<p>109 rows  36 columns</p>\n</div>"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print('grpm report:')\n",
    "display(pd.read_csv(db_path+'/GRPM_report.csv', index_col=0).T)\n",
    "print('grpm table genes:', len(pd.read_csv(db_path+'/grpm_table_output.csv').gene.drop_duplicates()))\n",
    "\n",
    "print('\\nnbib table:')\n",
    "display(pd.read_csv(db_path+'/complete_nbibtable.csv',index_col=0))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Show saved report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-03-14T10:52:32.906096400Z",
     "start_time": "2024-03-14T10:52:32.888095Z"
    }
   },
   "outputs": [],
   "source": [
    "# Visualize GRPM_report.csv\n",
    "GRPM_report = pd.read_csv(db_path+'/GRPM_report.csv', index_col=0).transpose().reset_index().rename(columns={'index':'gene'})\n",
    "\n",
    "repo_int_cols = ['lit2_variant', 'lit2_variant_norsid','lit2_rsid','lit2_rsid_plus1', 'lit1_rsid', 'lit1_rsid_pmid_plus1','lit1_pmid', 'lit1_pmid_pmid_plus1','pubmed_pmid_query',    'nbib_objects', 'nbib_objects_withdescriptors', 'pubmed_pmid', 'pubmed_pmid_withmesh', 'pubmed_pmidmesh','pubmed_mesh_qualifier_major','pubmed_mesh', 'rsid_pmid10','rsid_pmid50', 'rsid_pmid100' ]\n",
    "\n",
    "GRPM_report[repo_int_cols] = GRPM_report[repo_int_cols].astype(int)\n",
    "\n",
    "#display(GRPM_report_less.sort_values(by= 'matching_pmids',ascending=False))\n",
    "GRPM_report.sort_values(by='lit1_pmid',ascending = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2024-03-14T10:52:32.892115700Z"
    }
   },
   "outputs": [],
   "source": [
    "# Show Bar Graph\n",
    "GRPM_report_sort = GRPM_report.sort_values(by= 'pubmed_pmid',ascending=False)\n",
    "\n",
    "x = GRPM_report_sort.gene.iloc[:40]\n",
    "y = GRPM_report_sort['pubmed_pmid'].iloc[:40]\n",
    "plt.figure(figsize=(4, 8))\n",
    "plt.title('PMIDs in Dataset', loc='center',pad=10)\n",
    "\n",
    "plt.barh(x,y)\n",
    "plt.gca().invert_yaxis()\n",
    "plt.tick_params(axis='x', which='both', top=True, bottom=False, labeltop=True, labelbottom=False)\n",
    "#plt.xlabel('pmid count', position=(0.5, 1.08))\n",
    "plt.ylabel('genes')\n",
    "plt.xlabel('pmid count', position=(0.5, 1.08))\n",
    "ax = plt.gca()\n",
    "ax.xaxis.set_label_position('top')\n",
    "#plt.savefig('PMID_filtered.png',dpi=300, bbox_inches = \"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Debugging Code Block"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## debugging: litvar data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2024-03-14T10:52:32.895096Z"
    }
   },
   "outputs": [],
   "source": [
    "# if stucked check 'data' variable:\n",
    "# replace manually malformed lines\n",
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## debugging: nbib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2024-03-14T10:52:32.897093200Z"
    }
   },
   "outputs": [],
   "source": [
    "#NBIB PROBLEM SOLVER---------------------------------\n",
    "# replace malformed lines\n",
    "fullnbib= fullnbib.replace('2007/09/31','2007/09/30') # <= some dates are mispelled in pubmed\n",
    "ref = nbib.read(fullnbib)\n",
    "dfbib = pd.DataFrame(ref)\n",
    "dfbib"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "NBIB PROBLEM SOLVER [History] ---------------------------------\n",
    "with open('nbib report '+gene+'.txt', 'w', encoding='utf-8') as file:\n",
    "    file.write(fullnbib)\n",
    "with open('nbib report '+gene+'_FIXED.txt', 'r', encoding='utf-8') as file:\n",
    "    fullnbib = file.read() # --> not work"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# extras"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Eutils: get study type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2024-03-14T10:52:32.899094600Z"
    }
   },
   "outputs": [],
   "source": [
    "check_and_install_module('biopython')\n",
    "from Bio import Entrez\n",
    "Entrez.email = \"your_email@example.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2024-03-14T10:52:32.901094700Z"
    }
   },
   "outputs": [],
   "source": [
    "### EUTILS GET STUDY TYPE MODULE\n",
    "#https://biopython.org/docs/1.76/api/Bio.Entrez.html\n",
    "def get_study_type(pmids):\n",
    "    Entrez.email = 'your_email@your_domain.com'\n",
    "    handle = Entrez.esummary(db='pubmed', id=','.join(pmids), retmode='xml')\n",
    "    records = Entrez.parse(handle)\n",
    "    study_types = []\n",
    "    for record in records:\n",
    "        article_types = record['PubTypeList']\n",
    "        if 'Randomized Controlled Trial' in article_types:\n",
    "            study_types.append('Randomized Controlled Trial')\n",
    "        elif 'Controlled Clinical Trial' in article_types:\n",
    "            study_types.append('Controlled Clinical Trial')\n",
    "        elif 'Cohort Studies' in article_types:\n",
    "            study_types.append('Cohort Study')\n",
    "        elif 'Case-Control Studies' in article_types:\n",
    "            study_types.append('Case-Control Study')\n",
    "        elif 'Review' in article_types:\n",
    "            study_types.append('Review')\n",
    "        elif 'Clinical Trial' in article_types:\n",
    "            study_types.append('Clinical Trial')\n",
    "        elif 'Meta-Analysis' in article_types:\n",
    "            study_types.append('Meta-Analysis')\n",
    "        elif 'Multicenter Study' in article_types:\n",
    "            study_types.append('Multicenter Study')\n",
    "        else:\n",
    "            study_types.append('Unknown')\n",
    "    return study_types\n",
    "\n",
    "pmidlist = list(pmidmesh['pmids'].drop_duplicates())\n",
    "genepmids_str = list(map(str, pmidlist))\n",
    "study_type = get_study_type(genepmids_str)\n",
    "pmids_studytype = pd.DataFrame(list(zip(genepmids_str, study_type)), columns=[gene + '_PMID', 'study type'])\n",
    "request_counter += 1\n",
    "\n",
    "#study type count:\n",
    "pmids_studytype_count = pmids_studytype.groupby('study type').describe().reset_index()\n",
    "pmids_studytype_count.columns = pmids_studytype_count.columns.to_flat_index()\n",
    "new_column_names = ['study_type', 'pmid-count', 'pmid-unique', 'pmid-top', 'pmid-freq']\n",
    "pmids_studytype_count.columns = new_column_names\n",
    "pmids_studytype_countsort = pmids_studytype_count.sort_values(by='pmid-count', ascending=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
